### ISSUE TITLE
memory usage of streaming reduction

### ISSUE THREAD

**Issue by taobrienlbl** [LBL-EESA/TECA#402]
In #396, streaming mode was enabled for the table reduction stage in the Bayesian AR detector algorithm.  Analysis of the profiler output shows that the memory usage scales linearly with two factors: the number of threads, and the number of rows in the parameter table.  This limits the number of table rows that can be utilized: out-of-memory issues eventually occur.

Profiler output (see below) shows that memory usage increases linearly during the reduction stage. My initial hypothesis is that memory builds up because the reduction thread cannot keep up with the output generated by all of the working threads.

![Profiler output from running on 8192 table rows](https://user-images.githubusercontent.com/8796694/87091860-677d1900-c208-11ea-9943-87087d2c97b7.png)

We should analyze this in more depth to understand where the memory usage is coming from. If the above hypothesis is correct, @burlen suggested that we could multi-thread the reduction.

When we get around to testing this, a large version of the BARD parameter file, with a couple million rows, is available at NERSC: `/global/project/projectdirs/m1517/cascade/taobrien/teca_bayesian_ar_detector/mcmc_training/brute_force_analysis/teca_bruteforce_install/TECA/alg/teca_bayesian_ar_detect_parameters.cxx` (I would attach here, but it is over over 200MB.)

**Comment by burlen** [comment]
I agree that Travis's Hypothesis is a factor in the OOM condition. Multi-threading the reduction stage makes sense.

While there is an imbalance the capacity to generate and reduce the data, it may also be necessary to limit the number of threads generating data to be reduced. For instance one could set the the thread_pool_size to a multiplicative factor of the number of threads used in the reduction. Eg. 2 generator threads per reduction thread. This factor will be problem dependent.

Note that this could be done now on the command line. For instance in the above run, one could potentially avoid the OOM by reducing the number_of_threads to 16 instead of 32.

**Comment by burlen** [comment]
@taobrienlbl Could you help me reproduce this? I think having the large parameter table and path to IVT data would be enough.

**Comment by burlen** [comment]
All of the above assumes that a given rank has the resources to reduce a time step. when the table gets large enough it may not be the case. another approach would be to split the reduction (applied to each time step) over multiple MPI ranks. One could carve the communicator into groups of N where N ranks will cooperate to reduce each time step.

**Comment by burlen** [comment]
This should have been improved by bf99eff3
